content/toolsAndResources/nge/ngeBlocks.md

```diff
@@ -0,0 +1,1387 @@
+
+#### Properties
+- **Value** displays the value set in the node. To change the value, click on the plus (+) icon to get the input fields for each channel. These values are rounded to three decimal places by default though the node only displays two decimal places.
+
+#### Outputs
+- **value** is the Vector2 value set on the node.
+
+![Vector3 node](/img/tools/nge/vector3Node.jpg)
+
+### Vector3
+This input node represents a Vector3 value and contains the following properties which can be set in the properties panel. 
+
+#### Properties
+- **Value** displays the value set in the node. To change the value, click on the plus (+) icon to get the input fields for each channel. These values are rounded to three decimal places by default though the node only displays two decimal places.
+
+#### Outputs
+- **value** is the Vector3 value set on the node. 
+
+![Vector4 node](/img/tools/nge/vector4Node.jpg)
+
+### Vector4
+This input node represents a Vector4 value and contains the following properties which can be set in the properties panel. 
+
+#### Properties
+- **Value** displays the value set in the node. To change the value, click on the plus (+) icon to get the input fields for each channel. These values are rounded to three decimal places by default though the node only displays two decimal places.
+
+#### Outputs
+- **value** is the Vector4 value set on the node. 
+
+## Contextual
+Contextual nodes consist of values derived from the closest source associated with a contextual node in the graph. The way to determine what source is referenced by a contextual value is to trace wires from the contextual value node until finding a node that has a reference that could supply the context. For contextual nodes that reference geometric values like position or normal, trace wires to the first node that references geometry, and this will be the source that is being referenced.
+
+![Tracing contextual values to their source geometry](/img/tools/nge/tracingContext.jpg)
+
+In addition to geometric contextual values, there are other types of contextual values. One of them being LoopID which is a contextual value for the iteration of a node that loops. In this case, tracing wires to the first node that is iterable would be the source for the contextual value.  
+
+![Tracing non-geometric contextual values to their source](/img/tools/nge/tracingContext2.jpg)
+
+To help in identifying what source is feeding a contextual value, the graph will highlight the node providing the source of the contextual value when selecting any contextual value as can be seen in the images above. The playground below can be used to experiment with contextual values and better understand how they reference their source.
+
+<Playground id="#90J8BM#2" title="contextual values" description="Using contextual values to drive mesh creation." image="/img/playgroundsAndNMEs/contextualValuesPG.jpg"/>

```

Comment:
This PG does not work, there's an error in the console log:

```markdown
Uncaught TypeError: Cannot read properties of undefined (reading 'toString')
    at t.onreadystatechange (loadManager.ts:136:57)
```

Reply:
this may be due to changes in NGE as this example was created early and I've noticed several snippets breaking as the tool changes. I'm removing the example for now and will add one in later once it is more stable.

Reply:
removed for now in https://github.com/BabylonJS/Documentation/pull/963/commits/5395b92d4e3c0f28c82faf008026b09b521801ce